exporter/exportertest/contract_checker.go

```diff
@@ -0,0 +1,298 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package exportertest // import "go.opentelemetry.io/collector/exporter/exportertest"
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// UniqueIDAttrName is the attribute name that is used in log records/spans/datapoints as the unique identifier.
+const UniqueIDAttrName = "test_id"
+
+// UniqueIDAttrVal is the value type of the UniqueIDAttrName.
+type UniqueIDAttrVal string
+
+type CheckConsumeContractParams struct {
+	T *testing.T
+	// Factory that allows to create an exporter.
+	Factory exporter.Factory
+	// DataType to test for.
+	DataType component.DataType
+	// Config of the exporter to use.
+	Config               component.Config
+	NumberOfTestElements int
+	MockReceiver         interface{}
```

Comment:
This should be an interface providing the functionality needed for the contract checker only. `mockMetricsReceiver` and others should implement this interface in `exporter/otlpexporter` module. 

Probably better to do it thorough a factory
```suggestion
	MockReceiverFactory MockReceiverFactory
```

where 
```golang
type DecisionFunc func() error

type MockReceiverFactory func(DecisionFunc) MockReceiver

type MockReceiver interface {
     ... methods needed for contract checked
}

The main idea is that otlpexporter (or any other exporter) should depend on the contract checker, not the other way.

Reply:
Hey, I changed it `MockReceiver MockReceiver` for now, I think I'm missing the point of why and how to use `type MockReceiverFactory func(DecisionFunc) MockReceiver`. I keep the decision functions in contract checker to mimic the same behaviour that was implemented in receiver tests: https://github.com/open-telemetry/opentelemetry-collector/pull/7516/files#diff-a246dccd935c825bf517eb70b6476fdeb05f4750d960247b7df7fbf9a17b23aeR85
I would appreciate some more tips on how to use `MockReceiverFactory` ☺️

Reply:
The main goal is to move the actual test using the helper to the tested modules, e.g. `TestConsumeContractOtlpLogs` should go to `exporter/otlpexporter` module. Keeping that in mind should influence the right decisions for the `MockReceiver` interface. 

You need to pass DecisionFunc to the exporter test which should give you an implementation of `MockReceiver` in return. `MockReceiverFactory` should be implemented on the exporter side. Please let me know if it makes it any clearer.

Reply:
Ok, I think now it is more clear, I pushed another commit with MockReceiverFactory and tests moved under `exporter/otlpexproter`. It took some gymnastics with `net.Listen` (as now we create an instance every time depending on decision func). Please take a look and tell me if is it more or less what you meant.

Reply:
We are going in the right direction. Half way through :)